axon-server-connector/src/main/java/org/axonframework/axonserver/connector/event/axon/EventBuffer.java

```diff
@@ -139,44 +138,50 @@ public void blacklist(TrackedEventMessage<?> trackedEventMessage) {
 
     @Override
     public Optional<TrackedEventMessage<?>> peek() {
-        if (peekEvent == null && eventStream.hasNext()) {
-            peekEvent = eventStream.next();
-        }
-        return Optional.ofNullable(peekEvent);
+        return Optional.ofNullable(peekNullable());
     }
 
     @Override
     public boolean hasNextAvailable(int timeout, TimeUnit timeUnit) {
+        checkExceptionState();
         long deadline = System.currentTimeMillis() + timeUnit.toMillis(timeout);
         try {
-            do {
-                long waitTime = deadline - System.currentTimeMillis();
-                waitForData(waitTime);
-            } while (peekEvent == null && System.currentTimeMillis() < deadline && !eventStream.hasNext());
+            while (peekNullable() == null && System.currentTimeMillis() < deadline) {
+                lock.lock();
+                try {
+                    long waitTime = deadline - System.currentTimeMillis();
+                    // Check if an event has arrived or if the wait time is zero. In both cases, we don't have to wait.
+                    if (peekNullable() == null || waitTime <= 0) {
```

Comment:
Shouldn't this be `peekNullable() == null && waitTime > 0`

Reply:
Guess I was sleeping when I typed that...adjusted, thanks for calling it out.